File: impl/golang/internal/account/payment.go
package account

import (
	"fmt"

	"github.com/neuron-sdk/neuron-go-sdk/internal/keylib"
)

// FR-023: Parent payment address = evmAddress
// FR-024: Child payment address = parent evmAddress
// PaymentAddress resolves the payment address for this account.
//
// Resolution rules by account type:
//   - Parent: returns the account's own derived EVM address.
//   - Child: derives the Parent's EVM address from parentPubKey.
//   - Shared: returns an error because shared accounts have no single payment address.
func (a *NeuronAccount) PaymentAddress() (keylib.EVMAddress, error) {
	switch a.accountType {
	case Parent:
		if a.evmAddress == nil {
			return keylib.EVMAddress{}, fmt.Errorf("Parent account has no derived EVM address")
		}
		return *a.evmAddress, nil

	case Child:
		if a.parentPubKey == nil {
			return keylib.EVMAddress{}, fmt.Errorf("Child account has no parent public key")
		}
		parentAddr := a.parentPubKey.EVMAddress()
		return parentAddr, nil

	case Shared:
		return keylib.EVMAddress{}, fmt.Errorf("Shared accounts have no payment address")

	default:
		return keylib.EVMAddress{}, fmt.Errorf("unspecified account type has no payment address")
	}
}

File: impl/typescript/src/account/payment.ts
/**
 * Payment address resolution for NeuronAccount.
 *
 * Spec reference: 001 spec.md
 *   - FR-023: Parent's payment address is its own EVM address.
 *   - FR-024: Child's payment address resolves to Parent's EVM address
 *     via parentPubKey derivation.
 *   - SC-011: Payment address resolution is deterministic and verifiable.
 *
 * Shared accounts do not have a payment address (no single public key).
 */

import type { EVMAddress } from '../keylib/evm-address.js';
import type { NeuronAccount } from './account.js';
import { isParent, isChild } from './account.js';
import { invalidAccountType } from './errors.js';

/**
 * Resolve the payment address for a NeuronAccount.
 *
 * FR-023: Parent accounts return their own EVM address (derived from publicKey).
 * FR-024: Child accounts return the Parent's EVM address (derived from parentPubKey).
 * Shared accounts do not have a payment address and throw.
 *
 * @param account - NeuronAccount to resolve payment address for
 * @returns EVMAddress for payment purposes
 * @throws AccountError NEURON-ACCT-001 if account is Shared
 */
export function paymentAddress(account: NeuronAccount): EVMAddress {
  if (isParent(account)) {
    return account.publicKey.evmAddress();
  }
  if (isChild(account)) {
    return account.parentPubKey.evmAddress();
  }
  throw invalidAccountType('Shared accounts do not have a payment address');
}
